### PR TITLE
Fix ALPN - more fixes

### PR DESCRIPTION
**Pull request by tmshort**
* Clear proposed, along with selected, before looking at ClientHello
* Save proposed whether or not there's a callback, which may change
* Add test case for above
* Clear NPN seen after selecting ALPN
* Minor documentation updates

### CODE REVIEW

File: ssl/t1_lib.c [hunk 1]
```diff
@@ -2550,6 +2549,10 @@ static int ssl_scan_serverhello_tlsext(SSL *s, PACKET *pkt, int *al)
                 return 0;
             }
             s->s3->alpn_selected_len = len;
+#ifndef OPENSSL_NO_NEXTPROTONEG
```
Comment (ekasper) [inline ssl/t1_lib.c]: Um, I don't understand this change. It's the server that should choose one or the other, no?

**Comment by ekasper** [comment]
You say "Save proposed whether or not there's a callback" but I can't see that change?

File: ssl/t1_lib.c (hunk 1 above)
Comment (ekasper) [inline ssl/t1_lib.c]: And either way, isn't this fragile depending on extension ordering?

Comment (tmshort) [inline ssl/t1_lib.c]: Ugh... looks as though my branch was not up-to-date with master, and the other ALPN changes are not in there? :-/ Let me double-check this and update the PR.

Comment (tmshort) [inline ssl/t1_lib.c]: Wrong function, supposed to be clienthello_late.

**Comment by tmshort** [comment]
The callbacks checks had already been removed, that was a leftover comment from updating the 1.0.2 patch. The clearing of npn_seen has been moved to its proper spot in client_hello processing. This should be "good" now.

**Comment by ekasper** [comment]
+1

Can you update the commit message to remove the reference to callback checks?

**Comment by tmshort** [comment]
Updated commit comment.

**Comment by richsalz** [comment]
pushed thanks!

**Pull request closed without merging**